Code/Framework/AzCore/AzCore/JSON/rapidjson.h

```diff
@@ -8,52 +8,97 @@
+#error vanilla rapidjson was included before including <AzCore/JSON/rapidjson.h>
+#endif
+
+// Only override RAPIDJSON_ customization macros which are not already defined.
+// This allows the user to pre-define these before including O3DE if they want to use default values
+// or their own values.
+
+#if !defined(RAPIDJSON_ASSERT)
+#   if defined(AZ_ENABLE_TRACING)
+#       define RAPIDJSON_ASSERT(x) AZ_Assert(x, "Assert[rapidjson]: " #x)
+#   else
+#       define RAPIDJSON_ASSERT(x) (void)(0)
+#   endif
+#endif
+
+#if !defined(RAPIDJSON_STATIC_ASSERT)
+   #define RAPIDJSON_STATIC_ASSERT(x) static_assert(x, "Assert[rapidjson]: " #x)
+#endif
+
+#if !defined(RAPIDJSON_HAS_CXX11_TYPETRAITS)
+#   define RAPIDJSON_HAS_CXX11_TYPETRAITS 1
+#endif
+
+#if !defined(RAPIDJSON_HAS_CXX11_RVALUE_REFS)
+#   define RAPIDJSON_HAS_CXX11_RVALUE_REFS 1
+#endif
 
-#if defined(AZ_ENABLE_TRACING)
-#   define RAPIDJSON_ASSERT(x) AZ_Assert(x, "Assert[rapidjson]: " #x)
-#else
-#   define RAPIDJSON_ASSERT(x) (void)(0)
+// note for future maintainers:
+// RapidJSON allows you to customize the default allocator type for documents by overriding with RAPIDJSON_DEFAULT_ALLOCATOR.
+// However, customizing the default type of rapidjson::Document and rapidjson::Value in that way will cause incompatibilities
+// with external software libraries that use unmodified RapidJSON.   It is not thus recommended
+// to actually set the value of RAPIDJSON_DEFAULT_ALLOCATOR or redefine what a rapidjson::Document itself is.
+// The other way to get a 'custom' allocator installed in RapidJSON is to make your own Document/Value/Pointer type, since it's templated
+// on allocator type - for example, implement your own allocator and do
+// using MyDocument = rapidjson::GenericDocument<rapidjson::UTF8<char>, MyAllocator, MyAllocator> and then use MyDocument in your code
+// instead of RapidJSON::Document.  Note GenericDocument<> customized must use customized GenericValue<> and GenericPointer<>
+// or else you haven't acutally plugged into all the places where memory is allocated and will lose some.
+// 
+// Unfortunately, rapidjson::Pointer can't deal with these customized values or documents due to a bug in the code where functions
+// like Set/Get/Create use the default template parameter instead of deducing it:
+// ie, they are set up like somefunction<T>(GenericValue<char> val) which will only accept default GenericValue<char, CrtAllocator>.
+// correct form: somefunction<T>(GenericValue<char, T::AllocatorType> val, which would allow MyValue<char, MyAllocator> to be passed in.
+// thus, it is not recommended to use customized documents, values, or pointers either, (until that bug is fixed).
+
+// this leaves just one avenue for customization of memory of Rapidjson - overriding the macros RAPIDJSON_xxxx (new,delete, malloc, realloc..)
+
+// forward all allocations to our own allocators.  But preferrably without acutally having to drag our allocators into this header:
```

Comment:
Nitpick typo: acutally -> actually

Reply:
I have decided to install a spell checker extension. :)